Dockerfile

```diff
@@ -1,47 +1,206 @@
+	cd ../ && \
+	rm -rf ./Python-${PYVER}
+
+# Packages
 #
-USER odoo
-RUN pip3 install --user unittest-xml-reporting
+RUN apt install -y \
+	git \
+    curl \
+    nodejs \
+    npm \
+    postgresql \
+    postgresql-client \
+    printer-driver-cups-pdf \
+    python3-pkg-resources \
+    python3-setuptools \
+    unzip \
+    wkhtmltopdf \
+    x11vnc \
+    xfonts-75dpi \
+    xvfb
+
 
-## Download a compatible version of chromedriver
+
+# Odoo requires a non-standard build of wkhtmltopdf for many use cases
+# (including running without a local X display).
 #
-USER root
+ENV H2P_BASE https://github.com/wkhtmltopdf/packaging/releases/download
+ENV H2P_VER 0.12.6
+ENV H2P_REL 1
+ENV H2P_FILE wkhtmltox_${H2P_VER}-${H2P_REL}.bionic_amd64.deb
+ENV H2P_URI ${H2P_BASE}/${H2P_VER}-${H2P_REL}/${H2P_FILE}
+# wkhtmltopdf deb dependency (no longer in package manager)
+ENV LIBSSL_FILE libssl1.1_1.1.1f-1ubuntu2_amd64.deb
+ENV LIBSSL_URI http://security.ubuntu.com/ubuntu/pool/main/o/openssl/${LIBSSL_FILE}
+# Chrome deb
+ENV CHROME_FILE google-chrome-stable_current_amd64.deb
+ENV CHROME_URI https://dl.google.com/linux/direct/${CHROME_FILE}
+
+# Download and install all the deb packages manually
+# using apt install rather than dpkg -i so dependencies resolve
+RUN wget ${CHROME_URI} && \
+    apt install -y ./${CHROME_FILE} && \
+    rm ./${CHROME_FILE}
+
+RUN wget ${LIBSSL_URI} && \
+    apt install -y ./${LIBSSL_FILE} && \
+    rm ./${LIBSSL_FILE}
+
+RUN wget ${H2P_URI} && \
+    dpkg -i ./${H2P_FILE} && \
+    rm ./${H2P_FILE}
+
+## Install chromedriver
+#
+# Exports only persist in a single run directive, so we have to do this in one line.
+# Strip build version number on the first endpoint, as chromedriver version can lag behind slightly
+# This gives us the format for the latest compatible chromedriver in the second endpoint.
 RUN export CHROME_VER=$(echo $(google-chrome --version) | grep -oE '([[:digit:]]+\.)+' | sed 's/.$//') \
-    && export CHROMEDRIVER_VER=$(curl --location --fail --retry 3 https://chromedriver.storage.googleapis.com/LATEST_RELEASE_$CHROME_VER) \
-    && curl --silent --show-error --location --fail --retry 3 --output /bin/chromedriver_linux64.zip "https://chromedriver.storage.googleapis.com/$CHROMEDRIVER_VER/chromedriver_linux64.zip" \
-    && unzip -d /bin /bin/chromedriver_linux64.zip \
+    && export CHROMEDRIVER_VER=$(curl --location --fail --retry 3 https://googlechromelabs.github.io/chrome-for-testing/LATEST_RELEASE_$CHROME_VER) \
+    && curl --silent --show-error --location --fail --retry 3 --output /bin/chromedriver-linux64.zip "https://edgedl.me.gvt1.com/edgedl/chrome/chrome-for-testing/$CHROMEDRIVER_VER/linux64/chromedriver-linux64.zip" \
+    && unzip -d /bin /bin/chromedriver-linux64.zip \
+    && mv /bin/chromedriver-linux64/chromedriver /bin/chromedriver \
     && chmod a+x /bin/chromedriver
 
 
-# UDES Odoo snapshot
+# Odoo's barcode generation uses reportlab, which tends to misdetect
```

Comment:
Not sure if it's in scope, but at some point we should probably:

- check if this is still an issue
- if it is, find and link the official bug report so that we can check if it ever gets fixed

Reply:
It's a good point, i'll flag it in an issue

Reply:
The symlinks fail anyway, as the packages which add the source are no longer there.

I'm guessing this was added as 11.0 selenium tests compared pdf outputs which included barcodes. We don't have anything that does that in 14.0, so it's difficult to say whether it's necessary or not.

I think i'll comment out for now, and put this in a tech debt issue and we can re-visit if barcode pdfs are ever added again in 14.0

Reply:
https://taiga.unipart.digital/project/udes-x/issue/5152